### PR TITLE
Use CUTLASS for both `trans_a` and `trans_b` on Ampere

### PR DESCRIPTION
**Pull request by dfyz**
~~**Note that this PR *can't* be merged as is; see the end of the description.**~~ I implemented a workaround, so this PR should be safe to merge.

This PR is based on imoneoi@'s awesome work [here](https://github.com/tgale96/grouped_gemm/pull/6). The problem sorting code (which is only [needed](https://github.com/NVIDIA/cutlass/blob/main/media/docs/grouped_scheduler.md#potential-for-sorting-problems-to-reduce-imbalance) for variable-size `k`, i.e., in the backward pass) was copied from their PR verbatim. Everything else is structured pretty much the same conceptually, the only difference is that I tried to make the changes very minimal. Notably:
  * the original PR got rid of hardcoded threadblock/warp/instruction shapes, I left them as is (they can be changed in a separate PR, if necessary)
  * the original PR got rid of the cuBLAS fallback for Hopper, I left it intact (I think that it should be easy to make Hopper use CUTLASS too, but again, this can be done in a separate PR)

The only caveat is that vanilla CUTLASS has a bug I [described previously](https://github.com/tgale96/grouped_gemm/pull/6#issuecomment-2179455300) when handling `k=0` in a grouped GEMM. I added a test that reliably crashes the backward pass with a `CUDA error: an illegal memory access was encountered` error. The dimensions of the GEMM are larger than what is used in the other tests, but making them smaller unfortunately makes the crash go away, since the bug leads to an OOB read, which can go undetected without using `compute-sanitizer`.

I'm currently not sure how to work around this. The best option would be upstream CUTLASS landing my PR fixing the bug (and/or implementing an alternative fix), but I'm really not sure when this will happen. Looking at their opened PRs, I see that it might take weeks, or even months for a PR to be merged. :/

### CODE REVIEW

**Comment by tgale96** [comment]
Hi! Thanks for the PR! I will have a detailed look at it, but I am curious if you can just "squeeze" `k=0` problems from the batch for the backward pass? If `k=0`, the output just needs to be zero'd, so you can skip the compute and avoid the bug?

File: csrc/grouped_gemm.cu [hunk 1]
```diff
@@ -302,49 +342,61 @@ void GroupedGemm(torch::Tensor a,
...
+  }
 
   // NOTE: We support transposition through the 'trans_b' flag.
   TORCH_CHECK(a.is_contiguous());
   TORCH_CHECK(b.is_contiguous());
+  TORCH_CHECK(c.is_contiguous());
+
 
   // NOTE: Use cuBLAS for SM90 until CUTLASS supports SM90-optimized grouped-gemm.
 #if !defined(GROUPED_GEMM_DEVICE_CAPABILITY) || GROUPED_GEMM_DEVICE_CAPABILITY != 80
```
Comment (tgale96) [inline csrc/grouped_gemm.cu:382]: It looks like nothing in the CUTLASS path is Ampere (SM80) specific, so we could update these ifdefs?

Maybe we could keep the library using the cuBLAS path for now, and have a define that will enable the CUTLASS path so that some of the larger users can take it for a spin before we merge it to main?

File: csrc/grouped_gemm.cu [hunk 2]
```diff
@@ -84,12 +96,23 @@ torch::Tensor CopyToDevice(const std::vector<T> &x, const torch::Device &device)
   return out;
 }
 
-template <typename Gemm>
+template <typename T>
+static void ReorderArray(T* data, const std::vector<size_t>& indices) {
+    // For now, simply create a copy of the data and then copy over to the original.
+    std::vector<T> copy(indices.size());
```
Comment (tgale96) [inline csrc/grouped_gemm.cu:102]: nit: you could define copy = data, and then do the permutation from copy directly into data and skip the memcpy?

Same amount of data movement, just a little cleaner, I think?

**Comment by tgale96** [comment]
Sorry for the delay on the review :) This got lost in my email it seems.

File: csrc/grouped_gemm.cu (hunk 1 above)
Comment (dfyz) [reply csrc/grouped_gemm.cu:403]: I don't suppose there's anything Ampere specific, but I unfortunately don't have any H100s or similar to test on (hopefully this will change soon). Perhaps we can update the ifdefs it in a separate PR later?

Regarding the define: that sounds like a great idea, but I wasn't sure if you were proposing to introduce a define for non-Ampere (Hopper, etc.) only, or a define to enable the full CUTLASS path in general (so that you need to enable it explicitly even on Ampere). To be safe, I implemented the latter in 2285bb4170dbea77b8f48e04f1c021a676a844aa: the users should set a `GROUPED_GEMM_FULL_CUTLASS=1` environment variable if they want to enable CUTLASS for the backward pass (aka transposed cases).

If you have something different in mind, I can easily change it. :)

**Comment by dfyz** [comment]
> If k=0, the output just needs to be zero'd, so you can skip the compute and avoid the bug?

I would really prefer to have it fixed on the CUTLASS side (given that the fix is essentially a one-liner), but they also [suggest](https://github.com/NVIDIA/cutlass/pull/1593#issuecomment-2192481870) to skip the offending problems. In 398709f0645418ff9dffe1070a222d789ca8e3f1, I implemented a workaround along these lines, but I really hope that this will get fixed in CUTLASS one day because the workaround just complicates the code for no good reason. :/

> Sorry for the delay on the review

No worries at all! Thanks for taking the time to review it.

File: csrc/grouped_gemm.cu (hunk 1 above)
Comment (tgale96) [reply csrc/grouped_gemm.cu:403]: The GROUPED_GEMM_FULL_CUTLASS mode SGTM! Could we replace the other ifdef guards with this as well?

**Comment by tgale96** [comment]
@mvpatel2000 would you be willing to test this on H100 for us?

File: csrc/grouped_gemm.cu (hunk 1 above)
Comment (dfyz) [reply csrc/grouped_gemm.cu:403]: We could try, but I'd be very careful to not break anything accidentally, since testing that the correct code-path is chosen for all configurations is non-trivial. I currently just build the library in different configurations, run a simple test under `nsys profile`, and check what kernels are actually launched, which is quite error-prone.

Let me try to write down the current behavior of the library very explicitly to make sure I'm not missing anything:
  1. When `TORCH_CUDA_ARCH_LIST` is set, we *never* use CUTLASS, at least after 18f75979a3b9d8b20fe8b48fbf827a5d25992348. I think that this configuration is used for the PyPI builds, meaning you never use CUTLASS if you just `pip install grouped_gemm`.
  2. When `TORCH_CUDA_ARCH_LIST` is unset and we build for non-Ampere, CUTLASS is again never used.
  3. When `TORCH_CUDA_ARCH_LIST` is unset and we build for Ampere, CUTLASS is used for *the forward pass* only (i.e., `!trans_a && !trans_b`).

The current iteration of my PR modifies case `#3` like this:
  * [...] Additionally, iff `GROUPED_GEMM_FULL_CUTLASS` is set, CUTLASS is also used for the backward pass (i.e., `trans_a || trans_b`).

If we want to get rid of the other ifdefs and simplify things, I'd suggest the following:
  * Drop the `GROUPED_GEMM_DEVICE_CAPABILITY` define/environment variable. I think it's redundant when we also have `TORCH_CUDA_ARCH_LIST`.
  * Extend my PR to gate both forward and backward CUTLASS paths behind `#ifdef GROUPED_GEMM_FULL_CUTLASS` (or probably just `GROUPED_GEMM_CUTLASS`, since `...FULL...` is now implied).

This will technically break case `#2` because you now have to specify an additional environment variable (`GROUPED_GEMM_CUTLASS`) when building manually to enable CUTLASS. I don't think this is a huge problems because I believe that most users just use `pip install grouped_gemm` (case `#1`), which will continue to use cuBLAS for everything.

So, something like 18b2e8ee6e8758f36bfecd7afeebbf798aba7eef. I haven't tested this extensively, so I'm not pushing this to the PR branch just yet. Just want to make sure you're okay with this approach.

**Comment by dfyz** [comment]
> would you be willing to test this on H100 for us?

I actually got my hands on an H100, so I will be able at least run unit tests tomorrow. Any other testing would be much appreciated though!

**Comment by mvpatel2000** [comment]
Encountered the following running same model code with env flag set on:
```                                               │
[rank7]: │ /usr/lib/python3/dist-packages/torch/_tensor.py:525 in backward              │
[rank7]: │                                                                              │
[rank7]: │    522 │   │   │   │   create_graph=create_graph,                            │
[rank7]: │    523 │   │   │   │   inputs=inputs,                                        │
[rank7]: │    524 │   │   │   )                                                         │
[rank7]: │ ❱  525 │   │   torch.autograd.backward(                                      │
[rank7]: │    526 │   │   │   self, gradient, retain_graph, create_graph, inputs=inputs │
[rank7]: │    527 │   │   )                                                             │
[rank7]: │    528                                                                       │
[rank7]: │                                                                              │
[rank7]: │ /usr/lib/python3/dist-packages/torch/autograd/__init__.py:267 in backward    │
[rank7]: │                                                                              │
[rank7]: │   264 │   # The reason we repeat the same comment below is that              │
[rank7]: │   265 │   # some Python versions print out the first line of a multi-line fu │
[rank7]: │   266 │   # calls in the traceback and some print out the last line          │
[rank7]: │ ❱ 267 │   _engine_run_backward(                                              │
[rank7]: │   268 │   │   tensors,                                                       │
[rank7]: │   269 │   │   grad_tensors_,                                                 │
[rank7]: │   270 │   │   retain_graph,                                                  │
[rank7]: │                                                                              │
[rank7]: │ /usr/lib/python3/dist-packages/torch/autograd/graph.py:744 in                │
[rank7]: │ _engine_run_backward                                                         │
[rank7]: │                                                                              │
[rank7]: │   741 │   if attach_logging_hooks:                                           │
[rank7]: │   742 │   │   unregister_hooks = _register_logging_hooks_on_whole_graph(t_ou │
[rank7]: │   743 │   try:                                                               │
[rank7]: │ ❱ 744 │   │   return Variable._execution_engine.run_backward(  # Calls into  │
[rank7]: │   745 │   │   │   t_outputs, *args, **kwargs                                 │
[rank7]: │   746 │   │   )  # Calls into the C++ engine to run the backward pass        │
[rank7]: │   747 │   finally:                                                           │
[rank7]: │                                                                              │
[rank7]: │ /usr/lib/python3/dist-packages/torch/autograd/function.py:301 in apply       │
[rank7]: │                                                                              │
[rank7]: │   298 │   │   │   │   "of them."                                             │
[rank7]: │   299 │   │   │   )                                                          │
[rank7]: │   300 │   │   user_fn = vjp_fn if vjp_fn is not Function.vjp else backward_f │
[rank7]: │ ❱ 301 │   │   return user_fn(self, *args)                                    │
[rank7]: │   302 │                                                                      │
[rank7]: │   303 │   def apply_jvp(self, *args):                                        │
[rank7]: │   304 │   │   r"""                                                           │
[rank7]: │                                                                              │
[rank7]: │ /usr/lib/python3/dist-packages/torch/cuda/amp/autocast_mode.py:142 in        │
[rank7]: │ decorate_bwd                                                                 │
[rank7]: │                                                                              │
[rank7]: │   139 │   @functools.wraps(bwd)                                              │
[rank7]: │   140 │   def decorate_bwd(*args, **kwargs):                                 │
[rank7]: │   141 │   │   with autocast(enabled=args[0]._fwd_used_autocast, dtype=args[0 │
[rank7]: │ ❱ 142 │   │   │   return bwd(*args, **kwargs)                                │
[rank7]: │   143 │                                                                      │
[rank7]: │   144 │   return decorate_bwd                                                │
[rank7]: │   145                                                                        │
[rank7]: │                                                                              │
[rank7]: │ /usr/lib/python3/dist-packages/megablocks/layers/glu.py:146 in backward      │
[rank7]: │                                                                              │
[rank7]: │   143 │   │   │   │   op=activation_fn, out_shape=ctx.w1_out_shape, out_dtyp │
[rank7]: │   144 │   │                                                                  │
[rank7]: │   145 │   │   # Compute dw2 with recomputed activation_fn output.            │
[rank7]: │ ❱ 146 │   │   dw2 = gg.backend.gmm(                                          │
[rank7]: │   147 │   │   │   activation_fn_out, ddsd_out, batch_sizes, trans_a=True)    │
[rank7]: │   148 │   │                                                                  │
[rank7]: │   149 │   │   # Compute dactivation_fn_out.                                  │
[rank7]: │                                                                              │
[rank7]: │ /usr/lib/python3/dist-packages/grouped_gemm/backend.py:27 in gmm             │
[rank7]: │                                                                              │
[rank7]: │   24 def gmm(a, b, batch_sizes, trans_a=False, trans_b=False, c=None):       │
[rank7]: │   25 │   if c is None:                                                       │
[rank7]: │   26 │   │   c = _allocate_output(a, b, batch_sizes, trans_a, trans_b)       │
[rank7]: │ ❱ 27 │   backend.gmm(a, b, c, batch_sizes, trans_a, trans_b)                 │
[rank7]: │   28 │   return c                                                            │
[rank7]: │   29                                                                         │
[rank7]: │   30                                                                         │
[rank7]: ╰──────────────────────────────────────────────────────────────────────────────╯
[rank7]: IndexError: Dimension out of range (expected to be in range of [-2, 1], but got
[rank7]: 2)
```

**Comment by dfyz** [comment]
> Encountered the following running same model code with env flag set on

That was actually my bug in selecting the cuBLAS fallback, I missed that [`this #ifdef`](https://github.com/tgale96/grouped_gemm/pull/14/commits/2285bb4170dbea77b8f48e04f1c021a676a844aa#diff-bffec0006eb97fc116595ecdecac2dbf848af93a6d61247922a60e467c0fb615R357) should have been `... || GROUPED_GEMM_DEVICE_CAPABILITY != 80`, just as in a similar ifdef below. The current version of `setup.py` doesn't actually enable CUTLASS when `TORCH_CUDA_ARCH_LIST` is set, which I suppose happened in your case.

To sidestep this problem entirely, I pushed a streamlined version in 3be87fb350d7dca487e2688d86eda99ae369a31e that only has two simple modes irrespective of what `TORCH_CUDA_ARCH_LIST` is set to:
  * `GROUPED_GEMM_CUTLASS=1 pip install .` uses CUTLASS for both forward and backward passes
  * `GROUPED_GEMM_CUTLASS=0 pip install .` (or just not setting `GROUPED_GEMM_CUTLASS`) uses cuBLAS for both forward and backward passes

Notice that I changed `GROUPED_GEMM_*FULL_*CUTLASS` to just `GROUPED_GEMM_CUTLASS` to simplify things.

File: csrc/grouped_gemm.cu (hunk 1 above)
Comment (dfyz) [reply csrc/grouped_gemm.cu]: > So, something like https://github.com/tgale96/grouped_gemm/commit/18b2e8ee6e8758f36bfecd7afeebbf798aba7eef.

Okay, I just pushed exactly this to the PR branch. As [shown here](https://github.com/tgale96/grouped_gemm/pull/14#issuecomment-2209551575), the current logic is quite error-prone: it's really not obvious that the define that enables CUTLASS doesn't actually always enable CUTLASS. And I think that having just one CUTLASS define for both fwd and bwd also makes it harder to misconfigure things.

To summarize: `setup.py` is now configured via the `GROUPED_GEMM_CUTLASS` environment variable, which can be only set to
 * `1` (use CUTLASS everywhere)
 * any other value or not set at all (use cuBLAS everywhere)

Also, I checked that unit tests work just fine on H100 (with both `GROUPED_GEMM_CUTLASS=1` and `GROUPED_GEMM_CUTLASS=0`).

Does this sound good?

**Comment by mvpatel2000** [comment]
@dfyz latest seems slower with cutlass, does this match your benchmarks? 

Here is a forward pass with a MoE FFN using GLU with 4 experts per GPU using uniform routing where we see it is ~1.5x slower. Unfortunately, I don't have time to profile any deeper at the moment :( 

Baseline:
image

New Kernel:
image

**Comment by dfyz** [comment]
@mvpatel2000 I only ran benchmarks on an A100, but the H100 situation is not unexpected. Basically, as I said in the description of the PR, I initially wanted to make the minimal amount of changes to use CUTLASS everywhere. The motivation was not to make the kernels run faster, but to avoid GPU<->CPU synchronization points when training MoE models, which is only possible with CUTLASS (I will introduce these changes in a separate PR). In particular, this meant I didn't pay much attention to kernel performance on H100 when @tgale96 asked me to also enable CUTLASS on SM 9.0.

Since this PR is pretty far from minimal already, I guess that ship has sailed, though. :) So I dug a little further.

Here's a small benchmark I used to roughly estimate the performance of a full MoE training run using Mixtral 8x22B matrix sizes:
```python
import torch
import grouped_gemm as gg


if __name__ == '__main__':
    # Mixtral 8x22B sizes.
    M = 16384
    K = 4096
    N = 14336
    E = 8
    x = torch.rand(M, K, dtype=torch.bfloat16, device='cuda')
    w = torch.rand(E, K, N, dtype=torch.bfloat16, device='cuda')

    x.requires_grad_(True)
    w.requires_grad_(True)

    batch_sizes = torch.tensor([M//E]*E)

    with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CUDA]) as prof:
        for _ in range(30):
            out = gg.ops.gmm(x, w, batch_sizes)
            grad = out.sum().backward()

    torch.cuda.synchronize()
    prof.export_chrome_trace(f'gmm_trace.json')
```

On A100, after I just pushed 223cd55f1a7dfecad250c613da5fb7f795588a34 to get rid of weird hardcoded shapes (as @imoneoi's did in his branch), CUTLASS seems to be on par with cuBLAS.

|   | Time per iteration, ms | Perfetto trace |
| ------------- | ------------- | -------- |
| cuBLAS  |  **23.8** | [cleaned_a100_cublas_trace.json](https://github.com/user-attachments/files/16113206/cleaned_a100_cublas_trace.json)|
| CUTLASS, before 223cd55f1a7dfecad250c613da5fb7f795588a34 | **31.6** | [cleaned_a100_cutlass_old_trace.json](https://github.com/user-attachments/files/16113219/cleaned_a100_cutlass_old_trace.json) |
| CUTLASS, after 223cd55f1a7dfecad250c613da5fb7f795588a34| **23.9** | [cleaned_a100_cutlass_new_trace.json](https://github.com/user-attachments/files/16113220/cleaned_a100_cutlass_new_trace.json) |

On H100, the performance is indeed much worse (even worse than ~1.5x), and 223cd55f1a7dfecad250c613da5fb7f795588a34 doesn't help much.

|   | Time per iteration, ms | Perfetto trace |
| ------------- | ------------- | -------- |
| cuBLAS  |  **8.5** | [cleaned_h100_cublas_trace.json](https://github.com/user-attachments/files/16113241/cleaned_h100_cublas_trace.json) |
| CUTLASS, before 223cd55f1a7dfecad250c613da5fb7f795588a34 | **15.0** | [cleaned_h100_cutlass_old_trace.json](https://github.com/user-attachments/files/16113302/cleaned_h100_cutlass_old_trace.json) |
| CUTLASS, after 223cd55f1a7dfecad250c613da5fb7f795588a34| **13.8** | [cleaned_h100_cutlass_new_trace.json](https://github.com/user-attachments/files/16113305/cleaned_h100_cutlass_new_trace.json) |

Apparently, as @imoneoi [mentioned](https://github.com/NVIDIA/cutlass/issues/1280), we'd need a CUTLASS GEMM optimized specifically for SM 9.0. The problem is that, according to [this table](https://github.com/NVIDIA/cutlass/blob/main/media/docs/functionality.md#cutlass-3x-kernels) and [the official example](https://github.com/NVIDIA/cutlass/blob/main/examples/57_hopper_grouped_gemm/57_hopper_grouped_gemm.cu) requires using CUTLASS 3.x API instead of CUTLASS 2.x. This is non-trivial: in addition to changing the code, it requires bumping the CUTLASS version used in `third_party`.

@tgale96 What do you think of merging this PR as is (i.e., with "naive" H100 support), and then making H100 go brrr in a later PR (I'm willing to do this)? Since using CUTLASS now requires an explicit opt-in, I think this should be safe.

Alternatively, we can drop CUTLASS H100 support for now (keeping the current behavior), and then add H100 properly in a separate PR.

**Comment by tgale96** [comment]
I think its fine to merge this with the ifdef guards - it'll be functional on everything and off by default. Do you mind updating the README to explain the build option?

It's pretty annoying that CUTLASS completely changed the API for H100, but not unexpected :)

**Comment by tgale96** [comment]
And thank you @mvpatel2000 for testing this on H100!

File: setup.py [hunk 3]
```diff
@@ -4,24 +4,14 @@
 import torch
 from torch.utils.cpp_extension import BuildExtension, CUDAExtension
 
-if os.environ.get("TORCH_CUDA_ARCH_LIST"):
```
Comment (tgale96) [inline setup.py:7]: I think we needed this for some platforms - is there a reason to delete this?

**Comment by tgale96** [comment]
Hi! Sorry for the delay. I think everything looks great except there is some logic being removed from `setup.py` that I believe we need to support multi-platform builds / builds on platforms where no GPU is present.

**Comment by tgale96** [comment]
And yes, I think you can bump the version in `setup.py`. Thanks!

**Comment by dfyz** [comment]
> I think everything looks great except there is some logic being removed from `setup.py` that I believe we need to support multi-platform builds / builds on platforms where no GPU is present.

Ah right, sorry. I think it's the other way round, though: in the current iteration of this PR, multi-platforms builds / builds without GPU should work fine, but when a GPU is present and `TORCH_CUDA_ARCH_LIST` is unset, we build the library for more architectures than needed.

AFAIU, `device_capability` (set by the code I'd removed), was used for two different things:
  1. Host-side choice between `cuBLAS` and `CUTLASS` based on `-DGROUPED_GEMM_DEVICE_CAPABILITY=...` (and hence on `device_capability`). We got rid of `GROUPED_GEMM_DEVICE_CAPABILITY`, so I thought the `device_capability`-related stuff was also safe to remove.
  2. Selecting the compute capabilities to build for. This is what I missed: when `TORCH_CUDA_ARCH_LIST` was unset, the `device_capability` was also used to prune the number of compute capabilities based on the current GPU model (by setting `nvcc_flags`).

I just restored the old logic (without the `-DGROUPED_GEMM_DEVICE_CAPABILITY=...` part). Thanks for noticing!

**Comment by dfyz** [comment]
> And yes, I think you can bump the version in `setup.py`.

Bumped to `0.1.5`.

**Comment by tgale96** [comment]
Hi Ivan - sorry again for the delay! I just merged the PR. Thanks so much for the new feature! Very excited to play with it.

**Comment by tmm1** [comment]
> then making H100 go brrr in a later PR (I'm willing to do this)?

hey @dfyz, thanks for all your work here! curious if you did any more investigation into CUTLASS 3.x or H100 perf?

> This is non-trivial: in addition to changing the code, it requires bumping the CUTLASS version used in `third_party`.

I tried bumping third_party/cutlass to v3.7.0 and things still build because of their [backwards compat mode](https://github.com/NVIDIA/cutlass/blob/main/media/docs/cutlass_3x_backwards_compatibility.md). But the performance didn't change at all, so probably need to use the [new apis](https://github.com/NVIDIA/cutlass/blob/main/media/docs/gemm_api_3x.md)

**Comment by dfyz** [comment]
Hi @tmm1, proper H100 support is still on my to-do list, but I haven't got around to it yet. I think that using the new API is required to utilize TMA and WGMMA, so yeah, it's not surprising that the performance didn't change. The first steps should be really straightforward:
  1. Keep the current code for anything below Hopper, because why fix what's not broken. :) Thanks for confirming the current version works fine with CUTLASS 3.7, by the way.
  2. For Hopper specifically, write something similar to [this example](https://github.com/NVIDIA/cutlass/blob/main/examples/57_hopper_grouped_gemm/57_hopper_grouped_gemm.cu). It uses FP8, but I don't see anything inherently FP8-specific there.

This should definitely improve performance, however one thing that's bothering me is that we currently use a hardcoded set of (default) shapes everywhere. Thankfully, for Ampere the default shapes were enough to get cuBLAS-like performance for the matrix sizes I was interested in. However, I'm not sure this will hold for Hopper.

So, ideally, we should generate multiple template instantiations with different tile/cluster sizes, test them all, then choose the most performant one. I really hope this won't be needed for Mixtral-like matrix sizes, though (I'm mostly interested in those). :)

**Comment by tmm1** [comment]
> 2\. For Hopper specifically, write something similar to [this example](https://github.com/NVIDIA/cutlass/blob/main/examples/57_hopper_grouped_gemm/57_hopper_grouped_gemm.cu). It uses FP8, but I don't see anything inherently FP8-specific there.

Makes sense. I started trying to integrate it in https://github.com/tgale96/grouped_gemm/compare/tmm1:tmm1/cutlass-upgrade but there are a number of changes required for the new API.
